IA-3762 useLoadedData helper hook

msilva-broad opened the pull request:
- useLoadedData will handle typical data request mechanics and translate them to 'Loading', 'Ready' or 'Error' status outcomes in standardized `LoadedState<T>` result object.  Returns a tuple similar to useState hook.  See jsDoc and unit tests.
- part of work for Export Analysis Modal Typescript and Unit Tests

msilva-broad committed:
IA-3762 useLoadedData helper hook
- useLoadedData will handle typical data request mechanics and translate them to 'Loading', 'Ready' or 'Error' status outcomes in standardized LoadedState<T> result object.  Returns a tuple similar to useState hook.  See jsDoc and unit tests.
- part of work for Export Analysis Modal Typescript and Unit Tests

n8landolt commented on src/libs/ajax/loaded-data/useLoadedData.test.ts:
```diff
@@ -0,0 +1,182 @@
+    const hookRender = renderHook(() => useLoadedData<TestData>())
+    const hookResult1: UseLoadedDataResult<TestData> = hookRender.result.current
+    const updateData = hookResult1[1]
+    void updateData(async (): Promise<TestData> => {
```
Wrap in act(() => { void updateData... })

This will get rid of the junk in the console output of running the test.
There are two other spots I'll point out that could use the same fix.

msilva-broad replied:
ugg... I had fixed this but intelliJ did a strange thing that seamed to knock the file version back a few minutes.

msilva-broad replied:
fixed

n8landolt commented on src/libs/ajax/loaded-data/useLoadedData.test.ts:
```diff
@@ -0,0 +1,182 @@
+    const updateData = hookResult1[1]
+
+    // Act
+    void updateData(async (): Promise<TestData> => {
```
Wrap in act(()=>{})

msilva-broad replied:
fixed

n8landolt commented on src/libs/ajax/loaded-data/useLoadedData.test.ts:
```diff
@@ -0,0 +1,182 @@
+    const updateData = hookResult1[1]
+
+    // Act
+    void updateData(async (): Promise<TestData> => {
```
Wrap in act(()=>{})

msilva-broad replied:
fixed

n8landolt reviewed (changes requested):
I think it would be nice to get rid of the console junk by wrapping with act!

msilva-broad committed:
IA-3762 useLoadedData helper hook
- added act() wrapper in unit tests creating hook side-effects to prevent spamming jest test output with warnings.

msilva-broad commented:
> I think it would be nice to get rid of the console junk by wrapping with act!

yup, fixed up.  Resurrected the fix I had already done for it before intelliJ gobbled it up somehow.

nawatts commented on src/libs/ajax/loaded-data/useLoadedData.ts:
```diff
@@ -0,0 +1,65 @@
+        state: result
+      })
+    } catch (err: unknown) {
+      const error = isPromiseReturn((err as Response).text) ?
+        Error(await (err as Response).text()) :
+        err
```
Since `isPromiseReturn` calls the function, this won't work. The second `text` will error with "TypeError: Failed to execute 'text' on 'Response': body stream already read". Suggest checking `if (err instanceof Response) { ... }` instead.

nawatts replied:
Not sure if maybe there's a better way to extract errors from Responses. I'd expect most error responses to contain JSON, so this would produce an error with a message containing a JSON encoded string. I guess `dataCall` could handle error responses and rethrow an error though...

msilva-broad replied:
yes... this is meant to be an initial iterative solution.  I have it on my list to spend some time thinking more deeply about our error handling mechanics and related details.

nawatts commented on src/libs/type-utils/type-helpers.ts:
```diff
@@ -1,4 +1,32 @@
+  const isPromiseFn = (
+    !!f && (
+      (f as Function).constructor.name === 'AsyncFunction' ||
+          (typeof f === 'function' && isPromise(f()))
```
I think calling `f` here makes this function liable to produce some subtle bugs when `f` has side effects. Especially since this doesn't always call `f` (it doesn't call `async` functions).

msilva-broad replied:
good point... let me think about this a bit...

msilva-broad replied:
ok, removed the promise predicates and created an isFetchResponse predicate instead.

msilva-broad committed:
IA-3762 useLoadedData helper hook
- switched away from isPromiseReturn to use an isFetchResponse type predicate instead to drive error cleanup logic in useLoadedData

nawatts commented on src/libs/type-utils/type-helpers.ts:
```diff
@@ -1,4 +1,18 @@
+ * Type predicate that will check if variable (obj) is a Fetch Response
+ * @param obj
+ */
+export const isFetchResponse = (obj: unknown): obj is Response => {
```
To check if an object is a Response, why do we need something more complicated than `instanceof Response`?

msilva-broad replied:
I generally favor duck-typing... "if it quacks like a duck, it's a duck".  This follows the "Program to interfaces, not instantiations" design principle.  But I am open to counter-points...

nawatts replied:
If going with duck typing, all `useLoadedData` uses is the `text` method. Why also require `status` and `statusText` properties?

msilva-broad replied:
This was to compensate for not being able to safely check that text was a method that returned a promise without calling it... status and statusText seemed like a reasonable "surface" to test against.  Duck typing also makes mocking way easier... which also influenced these three props as being what is checked against for the type predicate.

msilva-broad committed:
Merge branch 'dev' into msilva/IA-3762_useLoadedData-hook
